### DISCUSSION TITLE
Public libraries within a private monorepo

### DISCUSSION THREAD

**Discussion by NicHaley** [vercel/turborepo#1331] [Help]
I am looking for some suggestions on how I might structure my project. I am using Turborepo to share UI and config between some libraries (npm packages), as well as a dashboard app.

Ideally, the libraries should be public, while everything else should be private.

The only way I can think to do this is with git-submodules and nested monorepos. Are there any special considerations for this use-case in Turborepo?

**Comment by oneezy** [comment]
@NicHaley I'm interested in this too. Did you ever get this working on your end?

Any pointers team?
@jaredpalmer @tknickman @nathanhammond @gsoltis @kdy1 @alexkirsz @mehulkar

**Comment by EarlGeorge** [comment]
all you need is to create private repo as monorepo and publish the only packages that you wish to be public..

**Comment by mehulkar** [comment]
@NicHaley workspaces that are git submodules is not a case that we test currently, but please try it out and see how it goes. One issue I can think of is that we [use git to figure out changed files][1]. It's possible that could not work with submodules. File an issue with a reproduction if you run into anything! 

[1]: https://github.com/vercel/turbo/blob/main/cli/internal/scm/git.go

**Comment by superdorvil** [comment]
Just here to say I am going to try something like this out on my own I am a freelancer and I want to have a structure where all my personal projects are in a monorepo, and all of my client projects are in a their own repos. I want UI components and basic utility library functions to be shared between my monorepo and all of my client projects. Is this possible to do? I am going to try it out and let you guys know how it works out. If the team at Vercel ever starts supporting something like this can you guys just drop a reply to let me know how it works out for you as well. 

Thanks.

**Comment by oneezy** [comment]
hey there @NicHaley @mehulkar @superdorvil ,

I've been working on something since my last comment and made some decent headway

**GOAL:**
Create a shareable UI package lib `(@oneezy/ui)` that can be:

**1.** worked on and demo'd in a public repo [(oneezy/ui)](https://github.com/oneezy/ui)
**2.** npm installed and used in other projects [(oneezy/site)](https://github.com/oneezy/site)
**3.** worked on in a public or private monorepo as a git submodule [(oneezy/projects)](https://github.com/oneezy/projects)

**BENEFITS:** 
- no longer have to copy/paste the same components over and over when using in other projects.
- if using `@oneezy/ui` and make updates/fixes/features, other projects will inherit them
- provides workflow for having a public repo inside a private repo or monorepo enabling you to take PR's from opensource community

---

**Here are 3 the repos that demo the 3 goals mentioned above:**

:white_check_mark: **1. oneezy/ui** https://github.com/oneezy/ui
this is the shareable ui lib (atomic design) and uses **sveltekit**, **histoire** and **tailwind** under the hood 
- all ui components live inside  `./src/lib` and use the `$ui` alias 
- ui components are imported like... `import { Hero } from '$ui'`
- when you add new components/fixes/updates you can run `pnpm package` to create a `package/` dir that's ready to be `npm publish`'d

:x: **2. oneezy/site** https://github.com/oneezy/site
this is a _broken_ test site that's intended to demo the `npm installed` version of `@oneezy/ui`. 
> _From what I can tell it's not working because `@oneezy/ui > package.json` isn't exporting the production paths properly. for example, it's looking for `./src/lib/index.js` (dev) instead of `./index.js` (production) ... a fix is in progress but if anyone has some input i'm all ears._
- all ui components live inside `node_modules/@oneezy/ui`
- ui components are imported like... `import { Hero } from '@oneezy/ui'`
- this was only created to demo the bare bones of what was needed to import `@oneezy/ui` from npm and also how to include `@oneezy/ui/app.css`, `@oneezy/ui/tailwind.config.cjs`, and `@oneezy/ui/postcss.config.cjs` into your project

:white_check_mark: **3. oneezy/projects** https://github.com/oneezy/projects
this is a barebones monorepo (turborepo) consisting of 2 apps:
- `apps/+ui` (this is the [oneezy/ui](https://github.com/oneezy/ui) git submodule repo)
- `apps/site` (this is the [oneezy/site](https://github.com/oneezy/site) repo but was copied/pasted and slightly modified to be a workspace)
> _**note:** 
the `apps/+ui` workspace imports components with the `$ui` alias 
the `apps/site` workspace imports components with `@oneezy/ui` alias_

To run a dev a server and automatically open up the `@oneezy/ui` and `@apps/site` apps:
1. `git clone --recursive https://github.com/oneezy/projects.git`
2. `cd projects` 
3. `pnpm i` 
4. `pnpm dev` 

--- 

I welcome any feedback or PR's w/ these guys! Not sure if I'm using the best practices here but am open to suggestions :)

**Reply by cm-datrics** [reply]
This is really interesting, have you tried deploying via vercel hooked up to GitHub? 

Intrigued to know if Vercel would automatically init the submodules etc to get the latest dependency

**Reply by oneezy** [reply]
sorry for the late response! yes, that's exactly how i used it (pushing to github and vercel builds). im sure the _automatic init_ is possible but haven't set it up. i took a long break on working on this stuff due to some family issues however ive recently been back at it..so i do plan on running a lot more tests soon

**Comment by JulianKingman** [comment]
Doing this on a small project, it feels like a super-power! I have a turborepo that I'm working on, and a 3rd party package forked and submodules into the packages folder. Updates are colocated, and at the end I'll have changes to submit as a PR upstream.

**Comment by candidia** [comment]
I personally use copybara to copy my public packages to their own repos. There are github actions that make the integration pretty easy.

**Reply by oneezy** [reply]
i just looked into it and it looked promising but it requires java and that was a big turn-off for me unfortunately
